tools/addrundata/json.go

```diff
@@ -0,0 +1,419 @@
+	jsall := make(jsonSuite)
+
+	for testdir, tc := range ts {
+		if !tc.existing.hasData {
+			errorf("Missing rundata: %s", testdir)
+			ok = false
+			continue
+		}
+
+		u := tc.existing.testUUID
+		jc := jsall[u]
+		if jc == nil {
+			jc = &jsonCase{}
+			jc.parsedData = tc.existing
+			jc.testDirs = make(map[string]string)
+			jsall[u] = jc
+		}
+
+		if !reflect.DeepEqual(tc.existing, jc.parsedData) {
+			errorf("Test UUID %s has inconsistent data at %s and %#v", u, testdir, jc.testDirs)
+			ok = false
+			continue
+		}
+
+		kind := testKind(testdir)
+		if !isTestKind(kind) {
+			kind = ""
+		}
+		reldir, err := filepath.Rel(rootdir, testdir)
+		if err != nil {
+			reldir = ""
+		}
+		jc.testDirs[kind] = reldir
+
+		sec := testSection(jc.testPlanID)
+		js := jp[sec]
+		if js == nil {
+			js = make(jsonSuite)
+			jp[sec] = js
+		}
+		js[u] = jc
+	}
+
+	return jp, ok
+}
+
+// testSection returns the test section (e.g. RT-1) part of the test plan ID
+// (e.g. RT-1.2).
+func testSection(testPlanID string) string {
+	i := strings.IndexRune(testPlanID, '.')
+	if i < 0 {
+		i = len(testPlanID)
+	}
+	return testPlanID[:i]
+}
+
+// jsonQuote quotes the string using JSON convention.
+func jsonQuote(s string) string {
+	data, err := json.Marshal(s)
+	if err != nil {
+		return strconv.Quote(s)
+	}
+	return string(data)
+}
+
+// jsonPlan maps from the test section (e.g. RT-1, TE-1) to a JSON test suite which
+// contains the test cases in that test section.
+type jsonPlan map[string]jsonSuite
+
+// empty creates a new JSON object representing an empty testplan.
+func (jp jsonPlan) empty() map[string]any {
+	const title = "Feature Profiles Test Plan"
+	return map[string]any{
+		"text": title,
+		"type": "testplan",
+		"li_attr": map[string]any{
+			"rel":          "testplan",
+			"title":        title,
+			"introduction": jsonQuote("https://github.com/openconfig/featureprofiles"),
+		},
+	}
+}
+
+// sortedKeys returns the keys in jsonPlan sorted in version order.
+func (jp jsonPlan) sortedKeys() []string {
+	var keys []string
+	for k := range jp {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return lessVersion(keys[i], keys[j])
+	})
+	return keys
+}
+
+// merge updates an existing "testplan" JSON object.
+func (jp jsonPlan) merge(o map[string]any) {
+	todos := make(jsonPlan)
+	for k, v := range jp {
+		todos[k] = v
+	}
+
+	// Update existing children first.
+	oldchildren, _ := o["children"].([]any) // Even if !ok, nil is fine.
+	var children []any
+
+	for _, child := range oldchildren {
+		o, sec, ok := childSuite(child)
+		if !ok {
+			children = append(children, child) // Passthrough mal-formed testsuites.
+			continue
+		}
+		js := jp[sec]
+		if js == nil {
+			children = append(children, child) // Passthrough JSON-only testsuites.
+			continue
+		}
+		js.merge(o)
+		children = append(children, o)
+		delete(todos, sec)
+	}
+
+	// Update the todos that were missing from the JSON.
+	for _, sec := range todos.sortedKeys() {
```

Comment:
Why do we need to sort the keys here? running `json.Marshal` sorts the keys already.

Reply:
First, these children are objects in a list, so they appear in the same order in the slice that we have to sort first.

Secondly, builtin sorting is done using string comparison ("TE-10.1" < "TE-2.1"), but the sorting we do here is version sort ("TE-2.1" < "TE-10.1").